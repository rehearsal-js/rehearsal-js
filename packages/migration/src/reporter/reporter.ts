import fs from 'fs';
import ts from 'typescript';
import winston from 'winston';
import json from './formatters/json';

export type ReportDiagnosticInfo = {
  file: string;
  code: number;
  message: string;
  start: number;
  length: number;
  nodeKind?: string;
  nodeText?: string;
};

export type Report = {
  version: string;
  diagnostics: ReportDiagnosticInfo[];
};

/**
 * Representation of diagnostic and migration report.
 */
export default class Reporter {
  readonly basePath: string;

  private report: Report;
  private logger?: winston.Logger;

  constructor(basePath = '', logger?: winston.Logger) {
    this.basePath = basePath;
    this.logger = logger?.child({ service: 'rehearsal-reporter' });
    this.report = {
      version: ts.version,
      diagnostics: [],
    };
  }

  getFileNames(): string[] {
    const fileNames: string[] = [];
    for (const diagnostic of this.report.diagnostics) {
      if (!fileNames.includes(diagnostic.file)) {
        fileNames.push(diagnostic.file);
      }
    }

    return fileNames;
  }

  getDiagnostics(fileName: string): ReportDiagnosticInfo[] {
    return this.report.diagnostics.filter((diagnostic) => diagnostic.file === fileName);
  }

  /**
   * Appends am information about provided diagnostic and related node to the report
   */
  addDiagnostic(diagnostic: ts.DiagnosticWithLocation, node?: ts.Node): void {
    this.report.diagnostics.push({
      file: diagnostic.file.fileName.replace(this.basePath, ''),
      code: diagnostic.code,
      message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '. '),
      start: diagnostic.start,
      length: diagnostic.length,
      nodeKind: node ? ts.SyntaxKind[node.kind] : undefined,
      nodeText: node?.getText(),
    });
  }

  /**
   * Prints the current report using provided formatter (ex. json, pull-request etc.)
   */
  print(file: string, formatter: (report: Report) => string): string {
    const report = formatter(this.report);

    if (file) {
      fs.writeFileSync(file, report);
    }

    return report;
  }

  /**
   * Saves the current report information to the file in JSON format
   */
  save(file: string): void {
    this.print(file, json);
    this.logger?.info(`Report saved to: ${file}.`);
  }

  load(file: string): Reporter {
    if (!fs.existsSync(file)) {
      this.logger?.error(`Report file not found: ${file}.`);
    }

    this.logger?.info(`Report file found: ${file}.`);
    const content = fs.readFileSync(file, 'utf-8');
    const report = JSON.parse(content);

    if (!Reporter.isReport(report)) {
      this.logger?.error(`Report not loaded: wrong file format`);
    }

    this.report = report;
    this.logger?.info(`Report loaded from file.`);

    return this;
  }

  private static isReport(report: any): report is Report {
    return report && report.version !== undefined && report.diagnostics !== undefined;
  }
}
