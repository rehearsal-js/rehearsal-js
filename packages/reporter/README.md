# @rehearsal/reporter

Terminal and File reporter
