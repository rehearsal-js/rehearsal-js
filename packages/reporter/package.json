{
  "name": "@rehearsal/reporter",
  "version": "0.0.28",
  "description": "Rehearsal Reporter",
  "keywords": [
    "rehearsal",
    "reporter",
    "typescript"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rehearsal-js/rehearsal-js.git"
  },
  "license": "BSD-2-Clause",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "yarn tsc -b",
    "lint": "yarn npm-run-all lint:*",
    "lint:eslint": "yarn eslint --fix . --ext .ts",
    "lint:tsc-src": "yarn tsc --noEmit",
    "prepare": "yarn build",
    "test": "yarn nyc mocha"
  },
  "dependencies": {
    "winston": "^3.6.0"
  },
  "devDependencies": {
    "@types/chai": "^4.3.0",
    "@types/debug": "^4.1.7",
    "@types/eslint": "^8.4.1",
    "@types/mocha": "^9.1.0",
    "chai": "^4.3.6",
    "mocha": "^10.0.0",
    "nyc": "^15.1.0"
  },
  "peerDependencies": {
    "typescript": ">4.7"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "volta": {
    "node": "16.16.0",
    "yarn": "1.22.19"
  }
}
