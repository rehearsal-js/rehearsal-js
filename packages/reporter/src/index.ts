export { Reporter } from './reporter';
export { jsonFormatter } from './formatters/json-formatter';
export { mdFormatter } from './formatters/md-formatter';
export { sarifFormatter } from './formatters/sarif-formatter';
export { sonarqubeFormatter } from './formatters/sonarqube-formatter';
export { normalizeFilePath } from './normalize-paths';

export type { Location, Report, ReportItem, ReportSummary, ReportFormatter } from './types';
