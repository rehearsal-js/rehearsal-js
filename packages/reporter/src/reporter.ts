import { existsSync, readFileSync, writeFileSync } from 'fs';
import { DiagnosticCategory, flattenDiagnosticMessageText, SyntaxKind } from 'typescript';
import {
  type ProcessedFile,
  type Report,
  type ReportFormatter,
  type ReportItem,
  type Location,
} from './types';
import type { DiagnosticWithLocation, Node } from 'typescript';
import type { Logger } from 'winston';

type ReporterMeta = {
  projectName: string;
  basePath: string;
  commandName: string;
  tsVersion: string;
};

/**
 * Representation of diagnostic and migration report.
 */
export class Reporter {
  readonly basePath: string;

  public report: Report;
  private logger?: Logger;

  constructor(meta: ReporterMeta, logger?: Logger) {
    const { projectName, basePath, commandName, tsVersion } = meta;

    this.basePath = basePath;
    this.logger = logger?.child({ service: 'rehearsal-reporter' });

    this.report = {
      summary: {
        projectName: projectName,
        tsVersion: tsVersion,
        timestamp: new Date().toLocaleString('en-US', {
          year: 'numeric',
          month: 'numeric',
          day: 'numeric',
          hourCycle: 'h24',
          hour: '2-digit',
          minute: '2-digit',
          second: '2-digit',
        }),
        basePath: basePath,
        commandName: commandName,
      },
      items: [],
    };
  }

  getFileNames(): string[] {
    return [...new Set(this.report.items.map((item) => item.analysisTarget))];
  }

  getItemsByAnalysisTarget(fileName: string): ReportItem[] {
    return this.report.items.filter((item) => item.analysisTarget === fileName);
  }

  /**
   * Appends an element to the summary
   */
  addSummary(key: string, value: unknown): void {
    this.report.summary[key] = value;
  }

  /**
   * Appends am information about provided diagnostic and related node to the report
   */
  addItem(
    diagnostic: DiagnosticWithLocation,
    files: { [fileName: string]: ProcessedFile },
    fixed: boolean,
    node?: Node,
    triggeringLocation?: Location,
    hint = '',
    helpUrl = ''
  ): void {
    this.report.items.push({
      analysisTarget: diagnostic.file.fileName,
      files,
      errorCode: diagnostic.code,
      category: DiagnosticCategory[diagnostic.category],
      message: flattenDiagnosticMessageText(diagnostic.messageText, '. '),
      hint: hint,
      fixed,
      nodeKind: node ? SyntaxKind[node.kind] : undefined,
      nodeText: node?.getText(),
      helpUrl,
      //Original node that triggers the error; may not be the node where the codefix happens
      //Location in type ProcessedFile is where the codefix happens or the message is added
      //The two locations in most cases will be the same, but can be different
      nodeLocation: triggeringLocation || undefined,
    });
  }

  /**
   * Prints the current report using provided formatter (ex. json, pull-request etc.)
   */
  print(file: string, formatter: ReportFormatter): string {
    const report = formatter(this.report);

    if (file) {
      writeFileSync(file, report);
    }

    return report;
  }

  /**
   * Saves the current report information to the file in simple JSON format
   * to be able to load it later with 'load' function
   */
  save(file: string): void {
    this.print(file, (report: Report): string => JSON.stringify(report, null, 2));
    this.logger?.info(`Report saved to: ${file}.`);
  }

  /**
   * Loads the report exported by function 'save' from the file
   */
  load(file: string): Reporter {
    if (!existsSync(file)) {
      this.logger?.error(`Report file not found: ${file}.`);
    }

    this.logger?.info(`Report file found: ${file}.`);
    const content = readFileSync(file, 'utf-8');
    const report: Report = JSON.parse(content);

    if (!Reporter.isReport(report)) {
      this.logger?.error(`Report not loaded: wrong file format`);
      return this;
    }

    this.report = report as Report;
    this.logger?.info(`Report loaded from file.`);

    return this;
  }

  private static isReport(report: Report): report is Report {
    return report && report.summary !== undefined && report.items !== undefined;
  }
}
