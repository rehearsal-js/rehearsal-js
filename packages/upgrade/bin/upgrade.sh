#!/usr/bin/env bash

node "$(dirname $0)/../dist/src/cli.js" "$@"
