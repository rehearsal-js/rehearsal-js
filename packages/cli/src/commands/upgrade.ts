#!/usr/bin/env node

import { mdFormatter, Reporter, sarifFormatter } from '@rehearsal/reporter';
import { upgrade } from '@rehearsal/upgrade';
import { Command, InvalidArgumentError } from 'commander';
import { compare } from 'compare-versions';
import { debug } from 'debug';
import { Listr } from 'listr2';
import { resolve } from 'path';
import { createLogger, format, transports } from 'winston';

import execa = require('execa');

import { generateReports, reportFormatter } from '../helpers/report';
import { UpgradeCommandContext, UpgradeCommandOptions } from '../types';
import {
  addDep,
  determineProjectName,
  getLatestTSVersion,
  getLockfilePath,
  getPathToBinary,
  git,
  gitCheckoutNewLocalBranch,
  gitCommit,
  gitIsRepoDirty,
  isValidSemver,
  parseCommaSeparatedList,
  timestamp,
} from '../utils';

const DEBUG_CALLBACK = debug('rehearsal:ts');

export const upgradeCommand = new Command();

function parseTsVersion(value: string): string {
  if (isValidSemver(value)) {
    return value;
  } else {
    throw new InvalidArgumentError(
      'The tsVersion specified is an invalid string. Please specify a valid version as n.n.n'
    );
  }
}

upgradeCommand
  .name('upgrade')
  .description('Upgrade typescript dev-dependency with compilation insights and auto-fix options')
  .argument('[basePath]', 'Path to directory contains tsconfig.json', '.')
  .option('-b, --build <beta|next|latest>', 'typescript build variant', 'beta')
  .option(
    '-t, --tsVersion <tsVersion>',
    'override the build variant by specifying the typescript compiler version as n.n.n',
    parseTsVersion
  )
  .option('-o, --outputPath <outputPath>', 'Reports output path', '.rehearsal')
  .option(
    '-r, --report <reportTypes>',
    'Report types separated by comma, e.g. -r json,sarif,md',
    parseCommaSeparatedList,
    []
  )
  .option('-d, --dryRun', `Don't commit any changes`, false)

  .action(async (basePath: string, options: UpgradeCommandOptions) => {
    const logger = createLogger({
      transports: [new transports.Console({ format: format.cli() })],
    });

    basePath = resolve(basePath);

    // WARN: is git dirty check and exit if dirty
    if (!options.dryRun) {
      const hasUncommittedFiles = await gitIsRepoDirty();
      if (hasUncommittedFiles) {
        logger.warn(
          'You have uncommitted files in your repo. Please commit or stash them as Rehearsal will reset your uncommitted changes.'
        );
        process.exit(0);
      }
    }

    const tscPath = await getPathToBinary('tsc');

    DEBUG_CALLBACK('tscPath', tscPath);
    DEBUG_CALLBACK('options %O', options);

    // grab the consuming apps project name
    const projectName = (await determineProjectName()) || '';

    const reporter = new Reporter(projectName, basePath, logger);

    const tasks = new Listr<UpgradeCommandContext>(
      [
        {
          title: 'Setting TypeScript version for rehearsal',
          task: (_ctx: UpgradeCommandContext, task): Listr =>
            task.newListr((parent) => [
              {
                title: `Fetching latest published typescript@${options.build}`,
                exitOnError: true,
                task: async (ctx, task) => {
                  if (!options.tsVersion) {
                    ctx.latestELRdBuild = await getLatestTSVersion(options.build);
                    task.title = `Latest typescript@${options.build} version is ${ctx.latestELRdBuild}`;
                  } else {
                    ctx.latestELRdBuild = options.tsVersion;
                    task.title = `Rehearsing with typescript version ${ctx.latestELRdBuild}`;
                  }
                },
              },
              {
                title: 'Fetching installed typescript version',
                task: async (ctx, task) => {
                  const { stdout } = await execa(tscPath, ['--version']);
                  // Version 4.5.0-beta
                  ctx.currentTSVersion = stdout.split(' ')[1];
                  task.title = `Currently on typescript version ${ctx.currentTSVersion}`;
                },
              },
              {
                title: 'Comparing TypeScript versions',
                task: async (ctx) => {
                  if (compare(ctx.latestELRdBuild, ctx.currentTSVersion, '>')) {
                    ctx.tsVersion = ctx.latestELRdBuild;
                    parent.title = `Rehearsing with typescript@${ctx.tsVersion}`;
                    reporter.addSummary('tsVersion', ctx.tsVersion);
                  } else {
                    parent.title = `This application is already on the latest version of TypeScript@${ctx.currentTSVersion}. Exiting.`;
                    // this is a master skip that will skip the remainder of the tasks
                    ctx.skip = true;
                  }
                },
              },
            ]),
        },
        {
          title: `Bumping TypeScript Dev-Dependency`,
          skip: (ctx): boolean => ctx.skip,
          task: (ctx: UpgradeCommandContext, task): Listr =>
            task.newListr(() => [
              {
                title: `Bumping TypeScript Dev-Dependency to typescript@${ctx.tsVersion}`,
                task: async (ctx: UpgradeCommandContext) => {
                  if (!options.dryRun) {
                    // there will be a diff so branch is created
                    await gitCheckoutNewLocalBranch(`${ctx.tsVersion}`);
                  }
                  await addDep([`typescript@${ctx.tsVersion}`], true);
                },
              },
              {
                title: `Committing Changes`,
                task: async (ctx: UpgradeCommandContext) => {
                  if (options.dryRun) {
                    task.skip('Skipping task because dryRun flag is set');
                  } else {
                    // eventually commit change
                    const lockFile = (await getLockfilePath()) || '';
                    await git.add(['package.json', lockFile]);
                    await gitCommit(
                      `bump typescript from ${ctx.currentTSVersion} to ${ctx.tsVersion}`
                    );
                  }
                  await addDep([`typescript@${ctx.tsVersion}`], true);
                },
              },
            ]),
        },
        {
          title: 'Checking for compilation errors',
          skip: (ctx): boolean => ctx.skip,
          task: (_ctx: UpgradeCommandContext, task): Listr =>
            task.newListr(
              () => [
                {
                  title: 'Building TypeScript',
                  task: async (_ctx, task) => {
                    try {
                      await execa(tscPath, ['-b']);

                      // if we should update package.json with typescript version and submit a PR
                      task.newListr(() => [
                        {
                          title: 'Creating Pull Request',
                          task: async (ctx, task) => {
                            if (options.dryRun) {
                              ctx.skip = true;
                              task.skip('Skipping task because dryRun flag is set');
                            }
                            // the diff has been added and commited to the branch
                            // do PR work here and push branch upstream
                            task.title = `Pull-Request Link "https://github.com/foo/foo/pull/00000"`;
                            // on success skip the rest of the tasks
                            ctx.skip = true;
                          },
                        },
                      ]);
                    } catch (error) {
                      // catch the tsc build compliation error
                      // re-run with tsc --clean and try and mitigate the error with ts-migrate plugins
                      throw new Error('Compilation Error');
                    }
                  },
                },
              ],
              { exitOnError: false }
            ),
        },
        {
          title: 'Re-Building TypeScript with Clean',
          skip: (ctx): boolean => ctx.skip,
          task: async () => {
            await execa(tscPath, ['-b', '--clean']);
          },
        },
        {
          title: 'Fixing the source code',
          skip: (ctx): boolean => ctx.skip,
          task: async (ctx, task) => {
            const configName = 'tsconfig.json';

            await upgrade({ basePath, configName, reporter, logger });

            // TODO: Check if code actually been fixed
            task.title = 'Codefixes applied successfully';

            if (!options.dry_run) {
              // add everything to the git repo within the specified src_dir
              await git.add([`${options.basePath}`]);
              await gitCommit(`fixes tsc type errors with TypeScript@${ctx.tsVersion}`);
            }
          },
        },
      ],
      { exitOnError: false }
    );

    try {
      const startTime = timestamp(true);
      await tasks.run().then(async (ctx) => {
        DEBUG_CALLBACK('ctx %O', ctx);
      });
      logger.info(`Duration:              ${Math.floor(timestamp(true) - startTime)} sec`);

      const reportOutputPath = resolve(basePath, options.outputPath);
      generateReports(reporter, reportOutputPath, options.report, {
        json: reportFormatter,
        sarif: sarifFormatter,
        md: mdFormatter,
      });

      // after the reporter closes the stream reset git to the original state
      // need to be careful with this otherwise if a given test fails the git state will be lost
      // this is reset within the afterEach hook for testing
      if (options.dryRun) {
        // await git.reset(['--hard', '--', 'package.json', 'yarn.lock', `${flags.src_dir}`]);
        // await git(['restore', 'package.json', 'yarn.lock', `${flags.src_dir}`], process.cwd());
      }
    } catch (e) {
      logger.error(`${e}`);
    }

    return;
  });
