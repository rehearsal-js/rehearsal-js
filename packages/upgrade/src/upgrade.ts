import { dirname, resolve } from 'path';
import {
  DiagnosticFixPlugin,
  LintFixPlugin,
  ReRehearsePlugin,
  DiagnosticCheckPlugin,
  LintCheckPlugin,
} from '@rehearsal/plugins';
import { Reporter } from '@rehearsal/reporter';
import { RehearsalService } from '@rehearsal/service';
import { findConfigFile, parseJsonConfigFileContent, readConfigFile, sys } from 'typescript';
import { debug } from 'debug';
import type { Logger } from 'winston';

export type UpgradeInput = {
  basePath: string;
  configName?: string;
  reporter: Reporter;
  logger?: Logger;
};

export type UpgradeOutput = {
  basePath: string;
  configFile: string;
  sourceFiles: string[];
};

const DEBUG_CALLBACK = debug('rehearsal:upgrade');

/**
 * Provides semantic diagnostic information in @ts-expect-error comments and in a JSON report
 */
export async function upgrade(input: UpgradeInput): Promise<UpgradeOutput> {
  const basePath = resolve(input.basePath);
  const configName = input.configName || 'tsconfig.json';
  const reporter = input.reporter;
  const logger = input.logger;

  const plugins = [
    ReRehearsePlugin,
    LintFixPlugin,
    DiagnosticFixPlugin,
    LintFixPlugin,
    DiagnosticCheckPlugin,
    LintFixPlugin,
    LintCheckPlugin,
  ];

  DEBUG_CALLBACK('Upgrade started at Base path: %O', basePath);

  const configFile = findConfigFile(basePath, sys.fileExists, configName);

  if (!configFile) {
    const message = `Config file '${configName}' not found in '${basePath}'`;
    logger?.error(message);
    throw Error(message);
  }

  DEBUG_CALLBACK('Config file found: %O', configFile);

  const { config } = readConfigFile(configFile, sys.readFile);
  const { options, fileNames } = parseJsonConfigFileContent(
    config,
    sys,
    dirname(configFile),
    {},
    configFile
  );

  DEBUG_CALLBACK('Config file content: %O', { options, fileNames });

  const service = new RehearsalService(options, fileNames);

  for (const fileName of fileNames) {
    DEBUG_CALLBACK('Processing file: %O', fileName);

    let allChangedFiles: Set<string> = new Set();

    for (const pluginClass of plugins) {
      const plugin = new pluginClass(service, reporter, logger);
      const changedFiles = await plugin.run(fileName);
      allChangedFiles = new Set([...allChangedFiles, ...changedFiles]);
    }

    DEBUG_CALLBACK('Plugins Complete on: %O', fileName);

    // Save file to the filesystem
    allChangedFiles.forEach((file) => service.saveFile(file));
    DEBUG_CALLBACK('Files Saved: %O', allChangedFiles);
  }
  DEBUG_CALLBACK('Upgrade finished');

  return {
    basePath,
    configFile,
    sourceFiles: fileNames,
  };
}
