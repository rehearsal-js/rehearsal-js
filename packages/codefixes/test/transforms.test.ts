import { copyFileSync, readdirSync, readFileSync, mkdirSync, realpathSync, rmSync } from 'node:fs';
import { resolve, dirname, basename } from 'node:path';
import { fileURLToPath } from 'node:url';
import { copySync } from 'fs-extra/esm';
import { Reporter } from '@rehearsal/reporter';
import { afterEach, describe, expect, test } from 'vitest';
import { upgrade } from '@rehearsal/upgrade';
import { dirSync } from 'tmp';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('Test transform', function () {
  const fixturesDir = resolve(__dirname, 'fixtures');
  const codefixesDir = resolve(__dirname, '../src/fixes');

  // eg. ["addErrorTypeGuard","addMissingExport","addMissingTypesBasedOnInheritance","makeMemberOptional"]
  const transforms = readdirSync(codefixesDir);

  afterEach(() => {
    // cleanupTsFiles(fixturesDir, transforms);
  });

  test.each(transforms)('%s', async (transform) => {
    const codefixesFixesTransformFixtureDir = resolve(codefixesDir, transform, 'fixtures');

    copyFiles(transform, codefixesFixesTransformFixtureDir, fixturesDir);

    const upgradeProjectDir = resolve(fixturesDir, transform);

    await runUpgrade(upgradeProjectDir);

    // Compare each updated .ts file with expected .ts.output
    const expectedOutputFiles = getExpectedOutputFiles(codefixesFixesTransformFixtureDir);

    for (const expectedOutputFile of expectedOutputFiles) {
      const expectedOutput = readFileSync(expectedOutputFile).toString();
      const filename = basename(expectedOutputFile).replace('.output', '');
      const actualOutput = readFileSync(resolve(upgradeProjectDir, `${filename}`)).toString();

      expect(expectedOutput).toEqual(actualOutput);
    }
  });
});

/**
 * Compiles the project with LanguageService and apply autofix
 * on corresponding diagnostic messages
 */
async function runUpgrade(basePath: string): Promise<void> {
  const reporter = new Reporter({
    tsVersion: '',
    projectName: '@rehearsal/test',
    basePath: '',
    commandName: '@rehearsal/migrate',
  });
  await upgrade({ basePath, reporter, entrypoint: '' });
}

function getExpectedOutputFiles(path: string): string[] {
  const filesInput = readdirSync(path, { withFileTypes: true });
  // we only want files which end in .ts.output
  return filesInput
    .filter((file) => file.isFile() && file.name.endsWith('.ts.output'))
    .map((file) => resolve(path, file.name));
}

// we want to be sure we maintain the current formatting of the files
// eg we dont want prettier to run on the files which is why we copy the files as .input/.output
function copyFiles(
  transformName: string,
  fromTransformFixturesDir: string,
  toTestFixturesDir: string
): void {
  // get all files which match from the fromTransformDir and copyFileSync to the toTestFixturesDir
  const files = readdirSync(fromTransformFixturesDir, { withFileTypes: true });
  // stick the files here and create the dir
  const testFixtureTransformDir = resolve(toTestFixturesDir, transformName);
  mkdirSync(testFixtureTransformDir, { recursive: true });

  // we only want files which end in .ts.input
  files
    .filter((file) => file.isFile() && file.name.endsWith('.ts.input'))
    .forEach((file) => {
      // create the folder structure in the toTestFixturesDir
      // eg. /rehearsal-js/packages/upgrade/test/fixtures/addErrorTypeGuard
      // copy the file from the fromTransformFixturesDir to the toTestFixturesDir
      // copyFileSync(`${file}.input`, resolve(basePath, parse(file).base));
      // remove the .input from the filename
      const filename = file.name.replace('.input', '');
      copyFileSync(
        resolve(fromTransformFixturesDir, file.name),
        resolve(testFixtureTransformDir, filename)
      );
    });
}

function prepareTmpDir(dir: string): string {
  const migrateFixturesDir = resolve(__dirname, '../fixtures/app_for_migrate');
  const testSrcDir = resolve(migrateFixturesDir, 'src');
  const srcDir = resolve(testSrcDir, dir);
  const { name: targetDir } = dirSync();
  copySync(srcDir, targetDir);
  return realpathSync(targetDir);
}
