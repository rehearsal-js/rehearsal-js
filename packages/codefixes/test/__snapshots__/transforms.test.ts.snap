// Vitest Snapshot v1

exports[`Test transform > addErrorTypeGuard 1`] = `
"
// Case #1: try ... catch

const test = ['dummy'];

try {
  console.log('success');
} catch (e) {
  console.log((e as Error).message);
/* @ts-expect-error @rehearsal TODO TS18046: 'e' is of type 'unknown'. */
  console.log(e.notErrorMember);

  if ((e as Error).name === 'Test') {
    console.log((e as Error).name);
    console.log((e as Error)  .name);
    console.log((e as Error).  name);
    console.log((e as Error) . name);
  }

  if (test[1] === undefined && (e as Error).name === 'Test') {
    console.log((e as Error).name);
  }
}


// Case #2: not a try catch

function getObject<T>(): T {
  const object = ['dummy'];
  return object as unknown as T;
}

const dummy = getObject();

/* @ts-expect-error @rehearsal TODO TS18046: 'dummy' is of type 'unknown'. */
dummy.key;
"
`;

exports[`Test transform > addMissingTypesBasedOnInheritance 1`] = `
"// Basics

class Animal {
  title: string = '';

  test(test: string): string {
    return test;
  }

  say(message: string): string {
    return message;
  }

  feed(food: Food, quantity: number): boolean {
    console.log(food, quantity);
    return false;
  }
}

interface Wild {
  wild(place: string): string;
}

interface Happy {
  happy(num: number): void;
}

class Food {
  title: string;

  constructor(title: string) {
    this.title = title;
  }
}

class Dog extends Animal {
  say(message: string) : string {
    console.log(message);
    return 'bar';
  }
}

export class WildDog extends Dog implements Wild, Happy {
  say(message: string) : string {
    console.log(message);
    return 'bar';
  }

  override feed(food: Food, quantity: number) : boolean {
    console.log(food, quantity);
    return true;
  }

  wild(place: string) : string {
    return place;
  }

  happy(num: number) : void {
    console.log(num);
  }
}

export class Cat extends Animal {
  override say(text: string) : string {
    console.log(text);
    return 'meo';
  }

  feed(something: Food) : boolean {
    console.log(something);
    return false;
  }
}

// Generics

type Thing = {
  foo: number;
};

// Generic Interfaces

interface GenericThingInterface<A = number, B = string> {
  thing(a: A, b: B): void;
}

interface ConstrainedThingInterface<C extends Thing> {
  thing(c: C): void;
}

export class GenericThingImp1 implements GenericThingInterface {
  thing(a: number, b: string) : void {
    console.log(a, b);
  }
}

export class GenericThingImp2 implements GenericThingInterface<boolean, number> {
  thing(a: boolean, b: number) : void {
    console.log(a, b);
  }
}

export class GenericThingImp3 implements ConstrainedThingInterface<{ foo: number; bar: string }> {
  thing(c: { foo: number; bar: string; }) : void {
    console.log(c);
  }
}

// Generic Classes

class GenericThingClass<A = number, B = string> {
  thing(a: A, b: B): void  {
    console.log(a, b);
  }
}

class ConstrainedThingClass<C extends Thing> {
  thing(c: C): void  {
    console.log(c);
  }
}

export class GenericThingExt1 extends GenericThingClass {
  thing(a: number, b: string) : void {
    console.log(a, b);
  }
}

export class GenericThingExt2 extends GenericThingClass<boolean, number> {
  thing(a: boolean, b: number) : void {
    console.log(a, b);
  }
}

export class GenericThingExt3 extends ConstrainedThingClass<{ foo: number; bar: string }> {
  thing(c: { foo: number; bar: string; }) : void {
    console.log(c);
  }
}

// Multiple signatures

interface MultipleInterface {
  foo(a: number): number;
  foo(b: string): string;
  foo(a: number, b: string): boolean;

  moo(a: string): string;
}

class MultipleClass {
  boo(a: number): number;
  boo(b: string): string;
  boo(a: number, b: string): boolean;
  boo(ab: number | string, b?: string): number | string | boolean {
    return !ab || !b;
  }
}

export class MultipleImpl extends MultipleClass implements MultipleInterface {
/* @ts-expect-error @rehearsal TODO TS2416: Property 'boo' in type 'MultipleImpl' is not assignable to the same property in base type 'MultipleClass'..  Type '(ab: any, b?: any) => boolean' is not assignable to type '{ (a: number): number; (b: string): string; (a: number, b: string): boolean; }'..    Type 'boolean' is not assignable to type 'number'. */
  boo(ab, b?): boolean  {
    return !ab || !b;
  }

/* @ts-expect-error @rehearsal TODO TS2416: Property 'foo' in type 'MultipleImpl' is not assignable to the same property in base type 'MultipleInterface'..  Type '(ab: any, b?: any) => boolean' is not assignable to type '{ (a: number): number; (b: string): string; (a: number, b: string): boolean; }'..    Type 'boolean' is not assignable to type 'number'. */
  foo(ab, b?): boolean  {
    return !ab || !b;
  }

  moo(b: string) : string {
    return b;
  }
}
"
`;

exports[`Test transform > addMissingTypesBasedOnInlayHints 1`] = `
"export function test1(): string  {
  return 'Test this';
}

export const test2 = () => console.log('Test this')

export const test3 = () => 'Test this';


export function test4<Type>(arg: Array<Type>): Type[]  {
  return arg;
}


export async function test5(): Promise<TestClass>  {
  return new TestClass();
}

interface TestInterface {
  test1(): string | undefined;
}

class BaseTestClass {
  test2(param: boolean): number | string {
    return param ? 0 : '0';
  };
}

export class TestClass extends BaseTestClass implements TestInterface {
  test1() : string | undefined {
    return undefined;
  }

  test2() : string | number {
    return 0;
  }
}
"
`;

exports[`Test transform > makeMemberOptional 1`] = `
"export interface Animal {
  species: string;
  weight?: number;
}

export interface Car {
  make: string;
  model: string;
  color?: string;
}
"
`;

exports[`Test transform > makeMemberOptional 2`] = `
"export class Employee {
  name: string;
  badgeNumber?: string;

  constructor({name, badgeNumber} : { name: string; badgeNumber: string }) {
    this.name = name;
    this.badgeNumber = badgeNumber;
  }
}"
`;

exports[`Test transform > makeMemberOptional 3`] = `
"export interface Rabbit {
    color?: string;
    species: string;
    name: string;
}

export function feedRabbit(rabbit: Rabbit): void  {
    console.log(\`Rabbit \${rabbit.name} loves to eat fruit, not hay.\`);
}"
`;

exports[`Test transform > makeMemberOptional 4`] = `
"export default interface Human {
  height?: number;
  weight: number;
}"
`;

exports[`Test transform > makeMemberOptional 5`] = `
"import { Animal, Car } from './2790-import-1';
import { Employee } from './2790-import-2';
import * as RabbitStuff from './2790-import-3';
import Human from './2790-import-4';

const animal: Animal = {
    species: 'dog',
    weight: 90,
};

delete animal.weight;

const car: Car = {
    make: 'Ford',
    model: 'Focus',
    color: 'red',
};

delete car.color;

const employee = new Employee({name: 'Victor', badgeNumber: '12345'});
delete employee.badgeNumber;

const human: Human = {
  weight: 200,
  height: 180,
};
delete human.height;

const rabbit: RabbitStuff.Rabbit = {
  color: 'butterscotch',
  species: 'Holland Lop',
  name: 'Poca',
};
delete rabbit.color;

interface Person {
    name: string;
    age?: number;
    address?: string | null;
}

const person:Person = {
    name: 'Jane',
    age: 10,
    address: '123 Happy St'
};

delete person.address;

class Student{
    name: string;
    grade: number;
    gender?: string;

    constructor({name, grade, gender}: { name: string; grade: number; gender: string }) {
        this.name = name;
        this.grade = grade;
        this.gender = gender;
    }
}
const student = new Student({name: '', grade: 3, gender: 'male'});
delete student.gender;

type Tree = {
  species: string;
  age: number;
  height?: number;
}

const oak: Tree = {
  species: 'oak',
  age: 150,
  height: 40,
};
delete oak.height;"
`;
